src/Microsoft.OData.Core/ObjectPool.cs

```diff
@@ -0,0 +1,49 @@
+﻿//---------------------------------------------------------------------
+// <copyright file="ObjectPool.cs" company="Microsoft">
+//      Copyright (C) Microsoft Corporation. All rights reserved. See License.txt in the project root for license information.
+// </copyright>
+//---------------------------------------------------------------------
+
+using System;
+using System.Collections.Generic;
+
+namespace Microsoft.OData
+{
+    internal class ObjectPool<T>
+    {
+        private readonly List<T> _objectsInUse;
+        private readonly List<T> _objectsAvailable;
+        private readonly Func<T> _objectGenerator;
+
+        public ObjectPool(Func<T> objectGenerator)
+        {
+            _objectGenerator = objectGenerator ?? throw new ArgumentNullException(nameof(objectGenerator));
+            _objectsInUse = new List<T>();
+            _objectsAvailable = new List<T>();
+        }
+        public T GetObject()
+        {
+            if (_objectsAvailable.Count != 0)
+            {
+                var obj = _objectsAvailable[0];

```

Comment:
I believe because we remove the `0` element two lines down, so `0` is always the "next" available. It might make more sense to use a stack for this in that case, though I'm a bit skeptical of the .net stack implementation (no real reason, I just don't have enough experience with it to say that it's good). 

It might be best to just store the index of the next available, and increment/decrement from that. `List<T>.RemoveAt` is pretty expensive. I would guess you would have even more performance improvement using an index field instead.

Reply:
Changed implementation to use a fixed size array.

Reply:
Removed the ObjectPool and used the `DefaultObjectPool` from dotnet extensions.